[Tree]미션1 체스 구현 시작하기

choitree opened the pull request:
미션1 PR이 많이 지연되었습니다.

choitree committed:
pawn and pawntest commit (mission1)

choitree committed:
reset test code directory

choitree committed:
test : Add
add assertThat and change build gradle

choitree committed:
test : Add
add create_기본생성자() and make compile error

choitree committed:
fix : Solve compile error

make default constructor

choitree committed:
fix : Solve compile error

make default constructor

choitree committed:
feat : Do Mission2

Create Board and BoardTest

sonypark commented on src/main/java/net/tree/chess/Board.java:
```diff
@@ -0,0 +1,26 @@
+
+    public void add(Pawn pawnColor) {
+        pawnList.add(pawnColor);
+    }
```
타입은 Pawn인데 파라미터명은 pawnColor 이네요. 통일해주세요~

sonypark commented on src/main/java/net/tree/chess/Board.java:
```diff
@@ -0,0 +1,26 @@
+
+        Pawn pawn = pawnList.get(i);
+
+        return pawn;
```
1. 변수명, 파라미터명 등은 축양하지 말고 써주세요~ 
2. `findPawn()`는 현재 요구사항에는 필요없는 메서드인 것 같네요.

sonypark commented on src/test/java/net/tree.chess/BoardTest.java:
```diff
@@ -0,0 +1,37 @@
+
+
+public class BoardTest {
+
+    @Test
+    public void create() throws Exception {
+        Board board = new Board();
```
Exception이 필요할까요?
DisplayName을 사용하기로 하셨다면 여기도 써주세요~

choitree replied:
제가 미션1 커밋을 하고, PR은 미션2도 반영된 내용을 커밋한걸로 보여지는데 
혹시 알려주신 리뷰 내용을 여기에 반영해야될까요? 아니면 미션1 내용만으로 다시 수정해서 재차 PR 요청드려야 할까요?~~ 

git checkout mission1로 해서 idea .로 열면 pawn과 pawntest만 보여지는데 
PR 요청드리면서 미션2가 추가된 커밋으로 반영되었나봐요ㅠㅠ

choitree replied:
미션2의 90퍼센트정도가 현재 PR에 들어간 상태라서, 호눅스에게 확인 후 현재 주신 리뷰 기준으로 수정하라는 안내를 받았습니다. findPawn과 size 메소드도 그래서 포함된 상태입니다ㅠㅠ Exception도 마찬가지구요~~ 
이 부분을 감안 부탁드립니다.미션3부터는 정확하게 올리도록 하겠습니다!!

Exception은 미션2에 가이드로 올라온 코드를 가져온건데 혹시 가이드도 똑같이 적을 필요는 없는건가요?~~

sonypark replied:
가이드는 말 그대로 가이드일 뿐입니다. 가이드가 항상 맞지는 않습니다. `throws Exception` 은 해당 메서드에서 Exception을 발생시킬만한 로직이 있을 때 필요한 코드입니다. 현재 코드상에서는 예외를 발생시키는 로직이 없네요. 인텔리제이에서는 불필요한 코드인 경우 회색으로 표시를 해주니 참고하세요~ 마우스를 올려놓으면 친절하게 알림 메시지가 뜨기도 합니다:)

![image](https://user-images.githubusercontent.com/34808501/108502282-eb969380-72f5-11eb-822f-cc71e2da6900.png)

sonypark commented on src/test/java/net/tree.chess/BoardTest.java:
```diff
@@ -0,0 +1,37 @@
+
+        Pawn white = new Pawn(Pawn.WHITE_COLOR);
+        board.add(white);
+        assertEquals(1, board.size());
+        assertEquals(white, board.findPawn(0));
+
+
+        Pawn black = new Pawn(Pawn.BLACK_COLOR);
+        board.add(black);
+        assertEquals(2, board.size());
+        assertEquals(black, board.findPawn(1));
+
```
1. assertEquals보다 assertJ 라이브러리를 사용하는게 어떨까합니다. 테스트코드를 더 편리하게 작성할 수 있도록 해주는 라이브러리 입니다.(가독성 + 사용성)
2. [여기](https://github.com/codesquad-members-2021/java-chess/pull/23#discussion_r572844078) 참고하셔서 assertAll을 사용해보시는 걸 권장드립니다.

sonypark commented on src/test/java/net/tree.chess/BoardTest.java:
```diff
@@ -0,0 +1,37 @@
+        assertEquals(black, board.findPawn(1));
+
+    }
+
+    @DisplayName("컴파일 에러 발생시키기")
+    @Test
+    public void checkPawnList() throws Exception {
```
1. DisplayName은 해당 테스트가 어떤 테스트를 하는지 알기 쉽게 명시하는 용도입니다. 지금 메시지로는 왜 컴파일 에러를 발생시키는지 잘 모르겠네요. 좀 더 구체적으로 적어주시면 좋겠어요.
2. 여기도 Exception을 throw하는 코드가 있네요. 필요한 부분인가요~?

sonypark commented on src/test/java/net/tree/pieces/PawnTest.java:
```diff
@@ -0,0 +1,30 @@
+import static org.junit.Assert.*;
+
+public class PawnTest {
+
+    @Test
+    @DisplayName("흰색 폰이 생성되어야 한다")
+    public void create() {
+        verifyPawn("white");
+        verifyPawn("black");
```
@DisplayName("흰색 폰이 생성되어야 한다")과 테스트 내용이 맞지 않아요

choitree replied:
수정했습니다~

sonypark commented on src/test/java/net/tree/pieces/PawnTest.java:
```diff
@@ -0,0 +1,30 @@
+
+    void verifyPawn(final String color) {
+        Pawn pawn = new Pawn(color);
+        assertEquals(color, pawn.getColor());
+        assertThat(color ,equalTo(pawn.getColor()));
```
assert문은 통일 해주세요~

choitree replied:
수정했습니다~

sonypark commented on src/test/java/net/tree.chess/BoardTest.java:
```diff
@@ -0,0 +1,37 @@
+    @Test
+    public void checkPawnList() throws Exception {
+        Board board = new Board();
+        assertTrue(board.pawnList.add("7"));
+        assertTrue(board.pawnList.add(new Pawn()));
```
이 테스트는 돌아가나요? 체크해주세요~

++ 커밋하기 전에 테스트가 깨지는 곳은 없는지 확인하고 커밋 해주세요:)

choitree replied:
체스 판에 Pawn 이외의 객체가 추가되지 않도록 한다.
말의 리스트에 new Integer("7")과 같이 Pawn 이외의 객체를 추가해 컴파일 에러가 발생하는지 확인한다.
즉, Pawn 이외의 다른 객체를 추가하면 컴파일 에러가 발생하도록 구현해야 한다.

미션2의 4번 추가 테스트 부분(상기 내용)은 컴파일 에러가 발생할 수 밖에 없도록 코드를 구현해야하는데
그 상태로 PR을 요청하면 되는것인지 아니면 제가 확인하고 지우면 되는것인지 판단이 안서서 우선 기재했는데, 에러가 나면 확인하고 지우면 되는건가요??

sonypark replied:
네 컴파일 에러가 나는 것만 확인하고 지우시면 됩니다. 컴파일은 말 그대로 런타임 이전에 발생하는 에러이기 때문에 테스트케이스로 확인할 수 없습니다. 강의 자료에 있는 컴파일 에러가 나는지 확인하라는 부분은 정의한 타입 이외의 타입을 넣어보고 컴파일 에러가 나는지 직접 확인해보라는 의미인 것 같네요.

sonypark commented on src/test/java/net/tree/pieces/PawnTest.java:
```diff
@@ -0,0 +1,30 @@
+    @Test
+    public void create_기본생성자() throws Exception {
+        Pawn pawn = new Pawn();
+        assertEquals("white", pawn.getColor());
```
"white"라는 문자열보다 Pawn에 만들어 놓으신 자바 상수를 사용하는건 어떨까요?

sonypark reviewed (changes requested):
안녕하세요 Tree, 리뷰어 소니 입니다.😊
미션1 구현하느라 수고하셨습니다.
보완이 필요한 부분에 대해 코멘트 남겼으니 확인 후 반영부탁드립니다!

++ 커밋 메시지를 조금 더 구체적으로 작성해주시면 좋을 것 같습니다.

choitree committed:
fix : Fix mission1

미션2가 반영된 내용이라서, 리뷰주신 내용 기준으로 수정하다.
@DisplayName이 안나와서 build.gradle 다시 설정하다.
변수, 메소드 이름 수정하다.
assertAll로 테스트 수정하다.

choitree reviewed (commented):
리뷰내용으로 수정했습니다. 미션1인데 사실상 미션2의 내용이 거의 들어가있어서 알려주신 내용으로 우선 수정하였습니다. 다음꺼부터는 안꼬이게 잘 올려보도록 하겠습니다😭

choitree reviewed (commented):
푸시가 제대로 안되서  커밋이 중간에 들어가있네요ㅠㅠ
헷갈리실거 같아서 코멘트로 다시 남깁니다

sonypark commented on src/test/java/net/tree/pieces/PawnTest.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Test
+    @DisplayName("흰색, 검정색 pawn이 생성되어야 한다")
+    public void create() {
+        verifyPawn("white");
+        verifyPawn("black");
+    }
+
+    void verifyPawn(final String color) {
+        Pawn pawn = new Pawn(color);
+        assertEquals(color, pawn.getColor());
+
+    }
+
+    @Test
+    @DisplayName("기본생성자 만들기")
+    public void create_기본생성자() {
+        Pawn pawn = new Pawn();
+        assertEquals("white", pawn.getColor());
```
여기도 자바 상수를 이용해주세요

sonypark commented on src/main/java/net/tree/chess/Board.java:
```diff
@@ -0,0 +1,26 @@
+        return pawnList.size();
+    }
+
+    public Pawn findPawn(int i) {
+
```
구체적인 파라미터명으로 변경

sonypark reviewed (changes requested):
아직 반영 안 된 부분이 있는데 확인해주세요~

choitree committed:
fix : Fix mission1

다시 리뷰받은 내용 반영해서 커밋합니다.

choitree reviewed (commented):
일찍 잠들어서 지금 확인했습니다~~ 
리뷰를 받고 다시 커밋하면 이렇게 코멘트를 남기는 게 맞는지 모르겠지만
혹시 확인 못하실까봐 다시 코멘트 남깁니다😶
아 그리구, 첫 리뷰 넘넘 감사합니다!!

sonypark commented on build.gradle:
```diff
@@ -9,7 +9,9 @@ repositories {
 }
 
 dependencies {
+    implementation 'org.junit.jupiter:junit-jupiter:5.4.2'
     compile('ch.qos.logback:logback-classic:1.2.3')
-    testCompile("org.junit.jupiter:junit-jupiter-api:5.6.0")
     testCompile("org.assertj:assertj-core:3.11.1")
+    //testCompile 'junit:junit:4.12'
+    testCompile 'org.junit.jupiter:junit-jupiter:5.4.2'
 }
```
1. 주석은 지워주세요.
2. junit5 설정이 중복되어 들어가있네요. 하나는 지워주시고, test에서 필요한 의존성이니 `testImplementaion`으로 사용해주세요.

sonypark reviewed (approved):
첫 미션 진행하시느라 수고하셨습니다.
다음 미션도 화이팅입니다!💪

++gradle 설정 관련 코멘트는 다음 미션에 적용해주세요~
